polygeist/tools/cgeist/driver.cc

```diff
@@ -682,40 +698,46 @@ static int finalize(mlir::MLIRContext &context,
 }
 
 // Create and execute the MLIR transformations pipeline.
+template <typename ModuleTy>
 static int
-createAndExecutePassPipeline(mlir::MLIRContext &context,
-                             mlir::OwningOpRef<mlir::ModuleOp> &module,
-                             llvm::DataLayout &DL, llvm::Triple &triple, bool &LinkOMP) {
-  // MLIR canonicalization & cleanup.
-  int rc = canonicalize(context, module);
-  if (rc != 0) 
-    return rc;
-
-  // MLIR optimizations.
-  rc = optimize(context, module);
-  if (rc != 0) 
-    return rc;
-
-  // CUDA specific MLIR optimizations.
-  rc = optimizeCUDA(context, module);
-  if (rc != 0) 
-    return rc;
-
-  rc = finalize(context, module, DL, LinkOMP);
-  if (rc != 0) 
-    return rc;
+createAndExecutePassPipeline(mlir::MLIRContext &context, ModuleTy &module,
+                             llvm::DataLayout &DL, llvm::Triple &triple,
+                             bool &LinkOMP,
+                             llvm::ArrayRef<llvm::StringRef> FuncOpNames) {
+  for (const auto FuncOpName : FuncOpNames) {
```

Comment:
Here is a scenario I am thinking of:
   1. createAndExecutePassPipeline<mlir::ModuleOp>(...) is called to optimize the CPU module
   2. createAndExecutePassPipeline<mlir:GPUModuleOp>(...) is called to optimize the device module
  
The loop will run 4 times (assuming FuncOpName contains 2 elements in each invocation) ? Should'n the optimization pipeline run once for the CPU module and once for the GPU module ?  

I do not understand the reason we need to pass FuncOpNames. We could make `canonicalize`, `optimize` etc... templated functions (and specialize the template if necessary). Then here we would have something like:  

```
template <typename ModuleTy>
static int createAndExecutePassPipeline(mlir::MLIRContext &context, ModuleTy &module,
                             llvm::DataLayout &DL, llvm::Triple &triple,
                             bool &LinkOMP) {
   
   int rc = canonicalize<ModuleTy>(context, module);
   ...
   rc = optimize<ModuleTy>(context, module);
   ...
}
```

Reply:
I wasn't that familiar with MLIR's PassManager. Using nesting, we're running the pipeline once on each module now. Also, code looks cleaner.